[HDR] Use the ImageSource properties to detect whether an image has gain map or not

shallawa opened the pull request:
#### e419179f6f32cb9e195e789e64abc2aaf33fbd20
<pre>
[HDR] Use the ImageSource properties to detect whether an image has gain map or not
<a href="https://bugs.webkit.org/show_bug.cgi?id=291765">https://bugs.webkit.org/show_bug.cgi?id=291765</a>
<a href="https://rdar.apple.com/149565322">rdar://149565322</a>

Reviewed by Simon Fraser.

CGImageSource properties will be checked to know whether an image has gain
map or not. If the image has a gain map, the key { kCGImageSourceDecodeRequest,
kCGImageSourceDecodeToHDR } will be sent to ImageIO to decode the gain-map image.
Without it ImageIO will decode the SDR image only.

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::hasHDRGainMap const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::DecodingOptions):
(WebCore::DecodingOptions::shouldDecodeToHDR const):
(WebCore::DecodingOptions::isCompatibleWith const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::hasHDRGainMap const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldDecodeToHDR const):
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::appendImageSourceOptions):
(WebCore::appendImageSourceOption):
(WebCore::imageSourceOptions):
(WebCore::imageSourceThumbnailOptions):
(WebCore::ImageDecoderCG::hasHDRGainMap const):
(WebCore::ImageDecoderCG::createFrameImageAtIndex):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:

Canonical link: <a href="https://commits.webkit.org/293871@main">https://commits.webkit.org/293871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a74a68593bc776fc02ebfb27d11a66019dc62148

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100147 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105277 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28268 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76239 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33303 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103154 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90452 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56600 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15178 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8450 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50098 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85089 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8533 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107636 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27260 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85194 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86659 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84730 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21529 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29405 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21115 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27197 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32429 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27008 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30324 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28567 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ImageTypes.h:
```diff
@@ -68,6 +68,11 @@ enum class GammaAndColorProfileOption {
     Ignored
 };
 
+enum class DecodeToHDR {
```
Do we need both this and `DecodingFormat`?

shallawa replied:
No. This one is removed.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -378,6 +388,49 @@ EncodedDataStatus ImageDecoderCG::encodedDataStatus() const
+    if (!properties)
+        return false;
+
+    auto fileContentsProperties = (CFDictionaryRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyFileContentsDictionary);
```
Should we validate that the value is a CFDictionary?

shallawa replied:
Yes. Fixed.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -378,6 +388,49 @@ EncodedDataStatus ImageDecoderCG::encodedDataStatus() const
+    if (!fileContentsProperties)
+        return false;
+
+    auto imagesInfoArray = (CFArrayRef)CFDictionaryGetValue(fileContentsProperties, kCGImagePropertyImages);
```
Should we validate that the value is a CFArray? Same with all the ones below.

shallawa replied:
Yes. Fixed.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -378,6 +388,49 @@ EncodedDataStatus ImageDecoderCG::encodedDataStatus() const
+        if (!type)
+            continue;
+
+        if (kCFCompareEqualTo == CFStringCompare(type, kCGImageAuxiliaryDataTypeHDRGainMap, 0))
```
Generally we put the `kCFCompareEqualTo` on the right.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -58,6 +58,7 @@ class BitmapImageDescriptor {
     DestinationColorSpace colorSpace() const;
     std::optional<Color> singlePixelSolidColor() const;
     Headroom headroom() const;
+    bool hasHDRGainMapContent() const;
```
Maybe just `hasHDRGainMap() const` ?

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -88,12 +89,13 @@ class BitmapImageDescriptor {
         ColorSpace                  = 1 << 7,
         SinglePixelSolidColor       = 1 << 8,
         Headroom                    = 1 << 9,
+        HasHDRGainMapContent        = 1 << 10,
```
`HasHDRGainMap`

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -243,6 +243,11 @@ EncodedDataStatus BitmapImageSource::setData(FragmentedSharedBuffer* data, bool
 
+#if USE(CG)
+    // FIXME: Usually this is not needed. But currently this is the only way to get "correct" gain map data from ImageIO.
+    m_decoder = nullptr;
```
Does this imply that we fully re-decode on every incremental data chunk? What are the performance implications?

shallawa replied:
The documentation of `CGImageSourceUpdateData()` says:
```
/* Update the incremental image source `isrc' with new data.  The new data
 * must include all the previous data plus any additional new data. The
 * `final' parameter should be true when the final set of data is provided;
 * false otherwise. */
```
So I am not sure whether CG/ImageIO keeps the `CGImageSource` decoded metadata while calling `CGImageSourceUpdateData()` multiple times. If it they do, that means the CGImageSource metadata will to be recoded with every incremental update.

I do not know right now the perf implications of recreating the `CGImageSource` but I will investigate.

shallawa replied:
I turned it this is a bug below WebKit. I deleted this line and I filed rdar://149562544 to track it internally. So no perf testing is needed for now.

smfr commented on Source/WebCore/platform/graphics/DecodingOptions.h:
```diff
@@ -36,11 +36,17 @@ enum class DecodingMode : uint8_t {
     Asynchronous
 };
 
+enum class DecodeToHDR {
```
`: bool`

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/DecodingOptions.h:
```diff
@@ -69,9 +78,12 @@ class DecodingOptions {
         return sizeForDrawing()->maxDimension() >= other.sizeForDrawing()->maxDimension();
     }
 
+    DecodeToHDR decodeToHDR() const { return m_decodeToHDR; }
```
The name sounds imperative ("decode to HDR!") but maybe it's a `shouldDecodeToHDR()`.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/DecodingOptions.h:
```diff
@@ -69,9 +78,12 @@ class DecodingOptions {
 private:
     DecodingMode m_decodingMode;
     std::optional<IntSize> m_sizeForDrawing;
+    DecodeToHDR m_decodeToHDR;
```
It's probably better packed up next to `m_decodingMode` ?

shallawa replied:
Moved.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -127,27 +127,48 @@ static RetainPtr<CFMutableDictionaryRef> appendImageSourceOptions(RetainPtr<CFMu
     });
-    return appendImageSourceOptions(adoptCF(CFDictionaryCreateMutableCopy(nullptr, 0, options)), subsamplingLevel, sizeForDrawing);
+
+    auto extendedOptions = appendImageSourceOptions(adoptCF(CFDictionaryCreateMutableCopy(nullptr, 0, options)), subsamplingLevel, sizeForDrawing);
```
This code appears twice. Please share.

shallawa replied:
Duplication was removed.

shallawa committed:
[HDR] Use the ImageSource properties to detect whether an image has gain map or not
https://bugs.webkit.org/show_bug.cgi?id=291765
rdar://149565322

Reviewed by Simon Fraser.

CGImageSource properties will be checked to know whether an image has gain
map or not. If the image has a gain map, the key { kCGImageSourceDecodeRequest,
kCGImageSourceDecodeToHDR } will be sent to ImageIO to decode the gain-map image.
Without it ImageIO will decode the SDR image only.

* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::hasHDRGainMap const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::DecodingOptions):
(WebCore::DecodingOptions::shouldDecodeToHDR const):
(WebCore::DecodingOptions::isCompatibleWith const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::hasHDRGainMap const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::shouldDecodeToHDR const):
* Source/WebCore/platform/graphics/NativeImageSource.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::appendImageSourceOptions):
(WebCore::appendImageSourceOption):
(WebCore::imageSourceOptions):
(WebCore::imageSourceThumbnailOptions):
(WebCore::ImageDecoderCG::hasHDRGainMap const):
(WebCore::ImageDecoderCG::createFrameImageAtIndex):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:

Canonical link: https://commits.webkit.org/293871@main
